### ISSUE TITLE
Missing SPIFFS Files Causes Connection Issues

### ISSUE THREAD

**Issue by Braehead** [philbowles/esparto#21]
This didn't used to be an issue, but now:

If you fully wipe the device for upload, then it now fails to establish a WiFi connection.  The ESP diagnostics report a connection (by displaying IP address, etc.), but Esparto fails to accept this status.  Nor does it call the user WiFiConnected event.

The problem is cured by uploading the correct WebUI files from the project 'data' directory into SPIFFS to support the WebUI.

Now, if I was never using the WebUI, I never allocated SPIFFS space or left it empty and the software all worked fine.  You got a browser error if you tried to use the WebUI, but it worked fine without the WebUI files.

Esparto should be able to deal with the files being missing and operate normally if the coder desires (I believe personally).  It could report the missing files to the serial port to highlight the situation.  But not connecting when files are missing is a bit of an annoyance.

**Comment by philbowles** [comment]
I'm usually with you on these things...but I'll stand my ground on this one. There is little point connecting to WiFI if you can't actually use any WiFi facility. I guess you could argue that you might still want just MQTT...but not having a webUI is going to confuse people and generate support calls... etc etc.

This is not a bug or an issue - its a fundamental design decision. The only thing at issue is not having a plain English error message...but then - to be honest -  I couldn't envisage anybody wanting to do do this. It a build error, plain and simple. An error message for a one-in-a thousand build error is waste of heap.

Secondly, irrespective of whether you use WiFi or not, Esparto keeps its configuration data in SPIFFS so if you want to use Esparto at all. you need SPIFFS. AS I have mentioned before, its is fundamental to the fabric of Esparto and cannot be removed. I'm not sure what you have against allocating space that is of no use to anything else anyway. 

Given that you have to have SPIFFS, it's of little utility to deliberately avoid loading the mandatory SPIFFS with the data required for normal functioning. Again, I simply do not understand why you want to avoid it? It saves nothing, its absence causes problems and gives no benefit. What am I missing? Just think of it as much part of Esparto as the code.

How would Esparto "operate normally " ? Operating normally is responding to voice requests. To do requires a number of files.. not running a webserver immediately prevents any Amazon echo voice control. And REST...and any "mesh/grid" coming in 3.4 etc so what are you left with?

For one second let's call this "headless mode". Headless = no webUI that means no voice control. Unless you want to "chunk" up the options separately. a) Run a web server as an absolute prerequisite for voice, but deliberately don't have web UI? b) no server at all, REST control now lost too. c) allow rest but no UI and/or voice? Where should it end? The voice control needs files from SPIFFS. So now we have to have options in the config block for several complicated permutations, the code becomes horrendously complex (it's already hairy).

The only valid use case for this that I can think of is "I want Esparto (partially) to work even when I forget to upload it properly". Solution:

1.Allocate SPIFFS as per documentation
2. Load it with DATA folder if using WiFi

Unless I have missed something in a specific use case you have in mind it seems like MORE work to do this wrong than it is to do it the recommended way! :)

I'll leave this open for the one in a million chance you may yet convince me...but its a massive change, not a one'line fix.

**Comment by Braehead** [comment]
I agree with your side.

I actually always allocate/reserve SPIFFS space, so stating I don't was a red herring on my part.  I understand it holds a load of other bits to make Esparto 'work'.

I can see no point whatsoever in controlling functionality regarding SPIFFS contents, that is increasing the complexity somewhat.  However, there must be some way in telling the idiot user (me last night) that something is wrong with SPIFFS, preventing some (if not most) of the functionality of Esparto from working.  A nice curt serial message explaining the situation at hand would have saved me (and probably any other user rushing the programming), from hours of WTF state.

The unit I was testing with, taking in a data stream and passing it onward again was working fine.  The MQTT stuff had not been coded to 'tee' the processed stream off elsewhere.  When I erased and flashed, and could not connect to the WebUI looking at the POR issues, I then found no IP address displayed during OnConnect.  So I knew something was amiss, but it took too long to pinpoint the failure -> forgot to upload SPIFFS after compile/erase/program cycle!!

Would a teenie weenie error message be possible to point people in the right direction??

I would hope this is not a massive code change, after all the fact that you suppress the connection means you must have detected something was wrong in the first place.  So you're most of the way there.

**Comment by philbowles** [comment]
OOOOh you are SO persuasive! Hmmm may have to eat humble pie AGAIN! Looks like there may be a use case for "headless" mode...ESP-01 won't take > 64K SPIFFS AND the binary since 2.5.2. fattened everything up....

Do I REALLY want all that work so support obsolescent kit???

The error message is easy. Working in a "headless mode" needs more thinking.

**Comment by philbowles** [comment]
I'll keep this open for a while. I may suddenly have a flash of inspiration that makes implementing "headless" a "quick fix".

**Comment by philbowles** [comment]
This piece of code might amuse you:
```cpp
ESPArto::ESPArto(ESPARTO_CONFIG_BLOCK cb): AsyncWebServer(80) {
	Serial.begin(74880);
	if(SPIFFS.begin()){
		_Q.reserve(ESPARTO_Q_MAX); //
		for(auto const& c:cb) _config[c.first]=c.second;
		_readConfig();
		setConfigInt(ESPARTO_CHIP_ID,ESP.getChipId(),"%06X");
		if(!_configItemEmpty(ESPARTO_DEVICE_NAME)) SCIs(ESPARTO_DEVICE_NAME,string("ESPARTO-")+CIs(ESPARTO_CHIP_ID));		
		queueFunction(_synchroStart); // warning, Will Smith! no default context yet!!!	
	} //else Serial.println(F("FATAL: NO SPIFFS!"));
}
```

I mean its not like I hadn't already thought about it! :)

**Comment by philbowles** [comment]
OK you may have won....:) check THIS out: I made a few changes and compiled with no SPIFFS
```cpp
load 0x4010f000, len 1384, room 16 
tail 8
chksum 0x2d
csum 0x2d
v8b899c12
~ld
WARNING: NO SPIFFS!
_synchroStart
Mode: STA
PHY mode: N
Channel: 6
AP id: 0
Status: 1
Auto connect: 1
SSID (8): XXXXXXXX
Passphrase (8): XXXXXXXX
BSSID set: 0
CI:$10=20
CI:$100=The Beast
CI:$101=airtight
CI:$102=192.168.1.4
CI:$103=
CI:$104=1883
CI:$105=
CI:$106=XXXXXXXX
CI:$107=XXXXXXXX
CI:$108=admin
CI:$109=admin
CI:$11=0
CI:$110=lwt
CI:$112=2
CI:$113=time1.google.com
CI:$114=time2.google.com
CI:$12=1
CI:$13=0
CI:$200=1
CI:$202=17D848
CI:$208=3.3.1.1
0 _cancelFallback 00000000
0 _initiateFallbac⸮>⸮}}}}⸮                     _                _____  _____
| ____|___ _ __   __ _ _ __| |_ ___   __   _|___ / |___ /
|  _| / __| '_ \ / _` | '__| __/ _ \  \ \ / / |_ \   |_ \ 
| |___\__ \ |_) | (_| | |  | || (_) |  \ V / ___) | ___) |
|_____|___/ .__/ \__,_|_|   \__\___/    \_/ |____(_)____/
          |_| 
C:\Users\phil\Desktop\DEV\examples33\zz_fun\numberOfTheBeast\numberOfTheBeast.ino
SDK:2.2.1(cfd48f3)/Core:2.5.2=20502000/lwIP:STABLE-2_1_2_RELEASE/glue:1.1-7-g82abda3/BearSSL:a143020

TEMP 293 172523
lightSpeed 11000 173180
lightSpeed 11000 2750
0 T=283 GOT IP XXXXXXXX[GW:192.168.1.1] as 192.168.1.104 airtight
0 _cancelFallback 3fff1cd4
0 T=298 TRY _mqttConnect dev=airtight
0 MQTT: Connected as airtight to 192.168.1.4:1883 user= will=lwt msg=
16:57:54 RTC SYNC
16:58:02 CMD airtight/cmd/reboot[]
```

That last was in reaction to an MQTT command....

So actually, the "quick fix" may have just happened. What worries me thiugh is that this is of course with a device that has already worked. I'm seriously concerned re a beginner coming at this with a virgin device, and I really don't have time to test that at the moment...

I may have "accidentally" made the code far more robust than I thought!

**Comment by Braehead** [comment]
That code snippet was sort of what I had predicted ;)

My concern was to newbies (and stressed pros) and that classic missed step.  It happens all the time.  Keen as mustard and you just miss 1 line of instructions and wham! It doesn't work.  All deflated and sad, with not enough experience to get them out of the minor hole they are in.  Fault finding anything is a lost art these days and you & I are rare 'old school' type engineers.  I made (and still make) a good living at fault finding.

So there you are, what looked like an impossible puzzle has been rewardingly simples.
